lib/Sema/MiscDiagnostics.cpp

```diff
@@ -539,18 +545,25 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
 
       TC.diagnose(E->getStartLoc(), diag::value_of_metatype_type);
 
-      // Add fix-t to insert '()', unless this is a protocol metatype.
-      bool isProtocolMetatype = false;
+      // Add fix-t to insert '()', only if this is a metatype of
+      // non-existential type and has any 'init' memeber.
+      bool isExistential = false;
       if (auto metaTy = E->getType()->getAs<MetatypeType>())
-        isProtocolMetatype = metaTy->getInstanceType()->is<ProtocolType>();
-      if (!isProtocolMetatype) {
+        isExistential = metaTy->getInstanceType()->isAnyExistentialType();
+      if (!isExistential &&
+          !TC.lookupConstructors(const_cast<DeclContext *>(DC),
+                                 E->getType()).empty())
         TC.diagnose(E->getEndLoc(), diag::add_parens_to_type)
           .fixItInsertAfter(E->getEndLoc(), "()");
-      }
 
       // Add fix-it to insert ".self".
-      TC.diagnose(E->getEndLoc(), diag::add_self_to_type)
-        .fixItInsertAfter(E->getEndLoc(), ".self");
+      auto diag = TC.diagnose(E->getEndLoc(), diag::add_self_to_type);
+      if (E->canAppendCallParentheses())
+        diag.fixItInsertAfter(E->getEndLoc(), ".self");
+      else
+        diag
+          .fixItInsert(E->getStartLoc(), "(")
+          .fixItInsertAfter(E->getEndLoc(), ").self");
```

Comment:
Style nitpick: we usually don't bother chaining once the diagnostic has a name.

```
diag.fixItInsert(E->getStartLoc(), "(");
diag.fixItInsertAfter(E->getEndLoc(), ").self");
```

(and we'd probably use braces around the multi-line `else` either way, even though it's technically one statement this way, and then do the `if` branch as well for symmetry.)